# Лабораторная работа №3. NumPy

## Задание

Выполните следующие задания:

1. Откройте в файл в Google Colab (используйте собственный форк репозитория).
2. Решите задачи.
3. Сохраните результат в ваш репозиторий github в директорию ./les03
4. Создайте pull request в репозиторий https://github.com/chebotarevsa/dap-2024. Название pull request должно иметь формат "<Номер лабораторной работы>  <Номер группы> <ФИО>"
5. Сдайте работу в системе "Пегас", в отчет укажите ссылку на pull request

In [1]:
import numpy as np
np.version.version

'2.3.5'

1. Создать массив 10x10 со случайными значениями, найти минимум и максимум (низкая)

In [2]:
random_array = np.random.randint(0, 101, (10, 10))
minimum = random_array.min()
maximum = random_array.max()

print("Массив 10x10:")
print(random_array)

print("\nМинимум:", minimum)
print("Максимум:", maximum)

Массив 10x10:
[[12 52  1 49 36  3 10 87 13 72]
 [25 51  8 45 36 74 83 96 37 19]
 [31 97  1 43  4 78 51 18 31 73]
 [17 35  0 10 26 41 59 95  8 20]
 [84  5 11 85 40 34 88 34 71 84]
 [47 48 67 32  7 32 58 42  6 37]
 [87 26 54 53 35 84 35 49 23  3]
 [33 76 95 42 21  8 41 97  3 26]
 [68 75 27 66 63 22 76 17 60 92]
 [17 90 86 37 78 39 16 34 88 57]]

Минимум: 0
Максимум: 97


2. Создать 8x8 матрицу и заполнить её единицами и нулями в шахматном порядке

In [3]:
chessboard = np.array([[ (i + j) % 2 for j in range(8)] for i in range(8)])
print("Матрица 8x8 в шахматном порядке:")
print(chessboard)

Матрица 8x8 в шахматном порядке:
[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


3. Дан массив из 11 элементов, поменять знак у элементов, значения которых между 3 и 8

In [4]:
z = np.arange(11)
print(f'Исходная матрица: {z}')
z[(z > 3) & (z < 8)] *= -1
print(f'Измененная матрица: {z}')

Исходная матрица: [ 0  1  2  3  4  5  6  7  8  9 10]
Измененная матрица: [ 0  1  2  3 -4 -5 -6 -7  8  9 10]


4. Найти среднее заначение строки матрицы и вычесть его из каждого елемента строки

In [5]:
z = np.random.rand(5, 10)
print(f'Исходная матрица: \n{z}')
avg = z.mean(axis=1, keepdims=True)
print(f'Среднеарифметические значения: \n{avg}')
new_matrix = z - avg
print(f'Измененная матрица: \n{new_matrix}')

Исходная матрица: 
[[0.49665219 0.83174765 0.78965445 0.91976264 0.58074231 0.47251702
  0.83971355 0.03169537 0.07417246 0.70213121]
 [0.94752965 0.52546893 0.6961367  0.56782716 0.73222007 0.7967065
  0.8709236  0.99603093 0.20723898 0.88889381]
 [0.35492918 0.05876763 0.99006054 0.66939866 0.52958626 0.39922811
  0.49316076 0.48852043 0.44361763 0.97305069]
 [0.88974245 0.00753485 0.40420847 0.79475317 0.36992846 0.67453212
  0.34874014 0.89259369 0.11182372 0.67332556]
 [0.53600564 0.17713344 0.95592573 0.58696442 0.41691917 0.82603434
  0.72516657 0.41402412 0.89630284 0.36644519]]
Среднеарифметические значения: 
[[0.57387888]
 [0.72289763]
 [0.54003199]
 [0.51671826]
 [0.59009214]]
Измененная матрица: 
[[-0.07722669  0.25786876  0.21577557  0.34588376  0.00686342 -0.10136186
   0.26583466 -0.54218352 -0.49970642  0.12825232]
 [ 0.22463202 -0.1974287  -0.02676093 -0.15507048  0.00932244  0.07380887
   0.14802597  0.27313329 -0.51565865  0.16599618]
 [-0.18510281 -0.48126436  0.450

5. Поменяйте 1 и 2 строки матрицы

In [6]:
z = np.arange(25).reshape(5,5)
print(f'Исходная матрица: \n{z}')
z[[0, 1]] = z[[1, 0]]
print(f'Матрица после обмена: \n{z}')

Исходная матрица: 
[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]
Матрица после обмена: 
[[ 5  6  7  8  9]
 [ 0  1  2  3  4]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]


6. Дан массив 10x2 подеставлющий собой точки в декартовой системе координат, преобразовать в полярную

In [7]:
cartesian = np.random.randint(-10, 10, size=(10, 2))
print(f'Декартовы координаты:\n {cartesian}')

x = cartesian[:, 0]
y = cartesian[:, 1]
r = np.sqrt(x**2 + y**2)
theta = np.arctan2(y, x)

polar = np.column_stack((r, theta))
print(f'Полярные координаты:\n {polar}')

Декартовы координаты:
 [[ 5  2]
 [ 3 -3]
 [ 8 -8]
 [-1  5]
 [ 8  1]
 [ 2  9]
 [ 2 -5]
 [-4  2]
 [ 7  4]
 [-5  0]]
Полярные координаты:
 [[ 5.38516481  0.38050638]
 [ 4.24264069 -0.78539816]
 [11.3137085  -0.78539816]
 [ 5.09901951  1.76819189]
 [ 8.06225775  0.12435499]
 [ 9.21954446  1.35212738]
 [ 5.38516481 -1.19028995]
 [ 4.47213595  2.67794504]
 [ 8.06225775  0.51914611]
 [ 5.          3.14159265]]


7. Заменить максимальный элемент массива на ноль

In [8]:
arr = np.random.randint(1, 10, 5)
print(f'Исходный массив: {arr}')
max_index = np.argmax(arr)
arr[max_index] = 0
print(f'Измененный массив: {arr}')

Исходный массив: [2 1 9 1 1]
Измененный массив: [2 1 0 1 1]


8. Дан массив, добавить 1 к каждому элементу с индексом, заданным в другом массиве. Добавление производиться только 1 раз

In [9]:
size = 10
arr = np.random.randint(0, 100, size=size)
arr_index = np.random.choice(size, size=3, replace=False)
print(f'Исходный массив: {arr}')
print(f'Массив уникальных индексов: {arr_index}')
arr[arr_index] += 1
print(f'Массив после изменения: {arr}')

Исходный массив: [58 64 15 67 52 67 90 86 82 63]
Массив уникальных индексов: [1 0 7]
Массив после изменения: [59 65 15 67 52 67 90 87 82 63]


9. Дан вектор [1, 2, 3, 4, 5], построить новый вектор с тремя нулями между каждым значением

In [10]:
z = np.array([1,2,3,4,5])
print(f'Исходный вектор: {z}')
zeros = np.zeros(3)
new_v = np.array([])
for i in range(len(z)):
    new_v = np.append(new_v, z[i])
    if i < len(z) - 1:
        new_v = np.append(new_v, zeros)

print(f'Новый вектор: {new_v.astype(int)}')

Исходный вектор: [1 2 3 4 5]
Новый вектор: [1 0 0 0 2 0 0 0 3 0 0 0 4 0 0 0 5]


10. Найти ближайшее к заданному значению число в заданном массиве

In [13]:
arr = np.arange(100)
num = np.random.uniform(0,100)
print(f'Исходное чилсо: {num}')
print(f'Исходный массив: {arr}')
closest_index = np.argmin(np.abs(arr - num))
closest_value = arr[closest_index]
print(f'Ближайшее значение в массиве: {closest_value}')

Исходное чилсо: 14.814134268030287
Исходный массив: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]
Ближайшее значение в массиве: 15


## Вопросы




**1**. Какой тип (dtype) будут иметь элементы массива?
``` python
np.array([1, 4, 2, 5, 3]) - int64
np.array([1., 4, 2, 5, 3]) - float64
np.array([1, 4, 2, 5, 3], dtype='float128') - float128
np.array([1., 4, 2, 5, 3], dtype='complex128') - complex128
np.array([1, 4, 2, 5, 3], dtype ='float64') - float64
np.array([1, 4, 2, 5, 3], dtype =int64) - int64
np.array([1, 4, 2, 5, 3], dtype='bool') - bool
np.array([1, 4, 2, 5, 3], dtype='str') - U1 (или U2, зависит от длины строк)
np.array([1, 4, 2, 5, '3']) U1 (U2)
np.array([1, -4, 2, 0, 1j]) - complex128
```

**2**. Что будет выведено в консоль и почему?
+ np.array([1, 4, 2, 5, 3], dtype='str')[1] + '1':
  + *Результат:* '41'
  + *Почему:* Элемент с индексом 1 это строка '4', конкатенация с '1' дает '41'
+ np.array([1, -4, 2, 0, 3], dtype='bool')[3] or np.array([1, -4, 2, 0, 3])[1]:
  + *Результат:* -4
  + *Почему:* dtype='bool' преобразует числа: 0 в False, остальные в True. Элемент с индексом 3 равен 0 → False. Операция or возвращает первый истинный операнд или последний.
+ np.array([1, -4, 2, 0, 3], dtype='bool')[3] and np.array([1, -4, 2, 0, 3])[1]:
  + *Результат:* False
  + *Почему:* Элемент с индексом 3 равен 0 → False. Операция and возвращает первый ложный операнд
+ np.array([1, -4, 2, 0, True], dtype='int64')[-1]:
  + *Результат:* 1
  + *Почему:* Так как True преобразуется в 1
+ np.array([1, -4, 2, 0, 1j], dtype='int64')[1]:
  + *Результат:* TypeError: int() argument must be a string, a bytes-like object or a real number, not 'complex'
  + *Почему:* Комплексное число 1j нельзя преобразовать в int64.
+ np.array([1, -4, 2, 0, 1j], dtype='str')[-1]:
  + *Результат:* '1j'
  + *Почему:* Комплексное число преобразуется в строку '1j'
+ np.array([1, 4, 2, 5, '3'])[-1] * 3:
  + *Результат:* '333'
  + *Почему:* Умножение строки: '3' * 3 = '333'



**3**. Каков результат операции?

+ np.ones((3, 5), dtype=float)
  + *Результат:* Матрица 3X5 заполненная 1.
+ np.full((3, 5), 3.14)
  + *Результат:* Матрица 3X5 заполненная 3.14
+ np.linspace(0, 1, 5)
  + *Результат:* Массив от 0 до 5 с равномерным шагом
+ np.diag((1,3,5))
  + *Результат:* Матрица 3X3 с 1,3,5 элементами на диагонали
+ np.smile((1,3,5))
  + *Результат:* AttributeError: module 'numpy' has no attribute 'smile'
+ np.random.random((3, 3))
  + *Результат:* Матрица 3X3 со случайными числами от 0 до 1
+ np.eye(3)
  + *Результат:*  Матрица 3X3 с 1 на диагонали и 0 в оставшихся элементах
+ np.empty(3)
  + *Результат:* Массив на 3 элемента с "мусорными" данными из памяти



**4**. Дан двумерный массив X, что будет выведено в консоль?

``` python
X = np.array([[3, 5, 2, 4],
      [7, 6, 8, 8],
      [1, 6, 7, 7]])
```

+ X[0,0] - 3 (Элемент первой строки, первого столбца)
+ X[-1,-1] - 7 (Элемент последней строки, последнего столбца)
+ X[0:1,-1] - 4 (Первая строка, последний столбец)
+ X[0:2,-1] - [4, 8] (Первые две строки, последний столбец)
+ X[0:2,0:2] - [[3, 5], [7, 6]] (Первые две строки, первые два столбца)
+ X[1,1:3] - [6, 8] (Вторая строка, со 2-го по 3-й столбец (не включая 3-й индекс в Python))
+ X[1,:] - [7, 6, 8, 8] (Вся вторая строка)
+ X[:,:] - Весь массив
+ X[:] - Весь массив
+ X[::2,::2] - [[3, 2], [1, 7]] (Каждая вторая строка и каждый второй столбец)
+ X[0:3:2,0:3:2] - [[3, 2], [1, 7]]  (Строка с 0 по 3 с шагом 2, столбец с 0 по 3 с шагом 2)
+ X[::-1,::-1] - Обратный порядок массива по строкам и столбцам

**5**. Дан двумерный массив grid, в чем различее между операциями?

+ grid = np.arange(16).reshape((4, 4))
  + np.split(grid, [2]) -  Разделяет массив по строкам после индекса 2
  + np.split(grid, [2], axis=1) - Разделяет массив по столбцам после индекса 2


**6**. В чем различие между операциями?

+ [1, 2, 3] + [3, 2, 1] - конкатенация списков
+ np.array([1, 2, 3]) + np.array([3, 2, 1]) - поэлементное сложение массивов

**7**. Как в CodeLab измерить время выполнения кода?

В CodeLab измерить время выполнения кода можно с помощью стандартных библиотек Python. Два самых распространённых способа:
+ Использование модуля time
+ Использование модуля timeit

**8**. Дан массив X, каковы результаты операций?

M = np.array([[3, 5, 2, 4],[7, 6, 8, 8],[1, 6, 7, 7]])
+ np.sum(M) - Сумма всех элементов матрицы(64)
+ np.sum(M, axis=1) - Сумма элементов по строкам (для каждой строки): ([14, 29, 21])
+ np.sum(M, axis=0) - Сумма элементов по столбцам (для каждого столбца): ([11, 17, 17, 19])
+ np.sum(M, axis=2) - AxisError: axis 2 is out of bounds for array of dimension 2
+ np.sum(M, 1) - Аналогично np.sum(X, axis=1): ([14, 29, 21])
+ sum(M) - Для многомерных массивов sum(X) суммирует по первой оси (аналогично np.sum(X, axis=0) для списков, но для NumPy массивов — по строкам): ([11, 17, 17, 19])
+ sum(M,1) - Встроенная функция sum не поддерживает аргумент axis.

**9**. Дан массив X, что будет выведено в консоль?

X = np.array([1,2,3,4])
+ X > 3 - Поэлементное сравнение: ([False, False, False,  True])
+ X < 3 - Поэлементное сравнение: ([ True,  True, False, False])
+ X[X < 3] - Возвращает элементы, для которых условие X < 3 истинно: ([1, 2])
+ X[X > 3] - Возвращает элементы, для которых условие X > 3 истинно:([4])
+ X[X != 3] - Возвращает элементы, для которых условие X != 3 истинно: ([1, 2, 4])

**10**.  В чем преимущество массивов NumPy по сравнению с обычными списками.  Чем обусловлено это преимущество?

*1. Скорость выполнения операций*
+ Преимущество: Операции с массивами NumPy выполняются значительно быстрее, чем с обычными списками Python, особенно при работе с большими объёмами данных.
+ Обоснование: NumPy написан на C и использует оптимизированные алгоритмы для работы с данными.
Операции над массивами NumPy векторизированы — они выполняются на уровне компилированного кода, без необходимости использовать циклы Python.
Списки Python — это динамические структуры данных, и операции с ними требуют дополнительных накладных расходов (например, проверка типов, динамическое изменение размера).


*2. Эффективное использование памяти*
+ Преимущество: Массивы NumPy занимают меньше памяти, чем списки Python, для хранения тех же данных.
+ Обоснование: NumPy хранит данные в непрерывном блоке памяти и использует фиксированные типы данных (например, int32, float64), что уменьшает накладные расходы.
Списки Python хранят ссылки на объекты, а не сами данные, что увеличивает расход памяти.


*3. Удобство работы с многомерными данными*
+ Преимущество: NumPy предоставляет удобные инструменты для работы с многомерными массивами (тензорами), что сложно или невозможно реализовать с помощью вложенных списков.
+ Обоснование: NumPy поддерживает многомерные массивы (например, матрицы, тензоры) и операции над ними (например, матричное умножение, транспонирование).
Операции над многомерными массивами выполняются эффективно и лаконично.


*4. Богатый функционал для математических операций*
+ Преимущество: NumPy предоставляет обширный набор математических функций (линейная алгебра, статистика, преобразования Фурье и др.), которые оптимизированы для работы с массивами.
+ Обоснование: Библиотека включает функции для векторных и матричных операций, что упрощает написание математического кода.
Например, умножение матриц, вычисление определителя, решение систем линейных уравнений и т. д.


*5. Интеграция с другими библиотеками*
+ Преимущество: NumPy является основой для многих научных библиотек Python (например, Pandas, SciPy, Matplotlib, scikit-learn), что делает его стандартом для научных вычислений.
+ Обоснование: Большинство библиотек для анализа данных и машинного обучения ожидают данные в формате NumPy-массивов.
